### PR TITLE
Stats based join distribution type

### PR DESCRIPTION
**Pull request by arhimondr**
Resolves: https://github.com/prestodb/presto/issues/11745

### CODE REVIEW

**Comment by arhimondr** [comment]
Still working on additional tests

**Comment by arhimondr** [comment]
@sopel39 Plans for queries `tpcds/q02`, `tpcds/q05`, `tpcds/q54`, `tpcds/q71` changed. Plan are different because of UNION before join. The cost calculators didn't compute cost for `UNION`. Plans had changed after i applied the `Determine join distribution type based on stats `, as the new implementation only considers costs of the exchanges only, and not the cumulative cost. So even if the cumulative cost was lost on the UNION node, the new implementation was able to operate.

But since the culprit of the plan changes wasn't the `Determine join distribution type based on stats`, but the fact that we were loosing cumulative cost on a `UnionNode`, i also introduced a commit that fixes cost propagation over a `UnionNode`. After fixing the cost propagation, the `Determine join distribution type based on stats` doesn't change any plans.

**Comment by arhimondr** [comment]
CC: @picrinite (for some reason GitHub doesn't pull you when i'm trying to add you either as a reviewer or subscriber)

File: presto-main/src/main/java/com/facebook/presto/cost/ExchangeCostCalculator.java [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+
+        double cpuCost = probeSideSize + buildSideSize * numberOfNodesMultiplier;
+
+        if (replicated) {
+            // add the cost of a local repartitioning of build side copies
+            // cost of the repartitioning of a single data copy has been already added in calculateExchangeCost
+            cpuCost += buildSideSize * (numberOfNodesMultiplier - 1);
+        }
+
+        double memoryCost = buildSideSize * numberOfNodesMultiplier;
```
[the remark this answers is not in the document]
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/cost/ExchangeCostCalculator.java]: I think calling this class an `ExchangeCostCalculator` was a bad idea. I wanted to encapsulate all the `numberOfNodes` related estimates here. I couldn't come up with a better name. Any ideas what can be a better name?

File: presto-main/src/main/java/com/facebook/presto/server/CoordinatorModule.java [hunk 2]
```diff
@@ -222,13 +218,9 @@ protected void setup(Binder binder)
...
 
-        // statistics calculator
+        // stats and cost calculators
         binder.install(new StatsCalculatorModule());
-
-        // cost calculator
-        binder.bind(CostCalculator.class).to(CostCalculatorUsingExchanges.class).in(Scopes.SINGLETON);
-        binder.bind(CostCalculator.class).annotatedWith(EstimatedExchanges.class).to(CostCalculatorWithEstimatedExchanges.class).in(Scopes.SINGLETON);
-        binder.bind(CostComparator.class).in(Scopes.SINGLETON);
+        binder.install(new CostCalculatorModule());
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/server/CoordinatorModule.java]: Does it make sense to have `StatsCalculatorModule` and `CostCalculatorModule` as separate modules?
Maybe one would be enough and more convenient?

File: presto-main/src/main/java/com/facebook/presto/cost/CostCalculatorUsingExchanges.java [hunk 3]
```diff
@@ -226,7 +226,7 @@ public PlanNodeCostEstimate visitSpatialJoin(SpatialJoinNode node, Void context)
                     node,
                     node.getLeft(),
                     node.getRight(),
-                    Objects.equals(node.getDistributionType(), Optional.of(JoinNode.DistributionType.REPLICATED)));
+                    Objects.equals(node.getDistributionType(), SpatialJoinNode.DistributionType.REPLICATED));
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/cost/CostCalculatorUsingExchanges.java]: `node.getDistributionType() == SpatialJoinNode.DistributionType.REPLICATED`

File: presto-main/src/main/java/com/facebook/presto/cost/CostCalculatorWithEstimatedExchanges.java [hunk 4]
```diff
@@ -148,7 +148,7 @@ public PlanNodeCostEstimate visitSpatialJoin(SpatialJoinNode node, Void context)
             return calculateJoinCost(
                     node.getLeft(),
                     node.getRight(),
-                    Objects.equals(node.getDistributionType(), Optional.of(JoinNode.DistributionType.REPLICATED)));
+                    Objects.equals(node.getDistributionType(), SpatialJoinNode.DistributionType.REPLICATED));
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/cost/CostCalculatorWithEstimatedExchanges.java]: `node.getDistributionType() == SpatialJoinNode.DistributionType.REPLICATED`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java [hunk 5]
```diff
@@ -433,7 +437,7 @@ public PlanOptimizers(
                     ruleStats,
                     statsCalculator,
                     estimatedExchangesCostCalculator,
-                    ImmutableSet.of(new DetermineJoinDistributionType(costComparator))))); // Must run before AddExchanges
+                    ImmutableSet.of(new DetermineJoinDistributionType(costComparator, exchangeCostCalculator))))); // Must run before AddExchanges
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: Cost calculators are passed to Rules from IterativeOptimizer, via the Context

File: presto-main/src/main/java/com/facebook/presto/cost/CostCalculatorModule.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.cost;
+
+import com.google.inject.Binder;
+import com.google.inject.Module;
+import com.google.inject.Scopes;
+
+public class CostCalculatorModule
```
[the remark this answers is not in the document]
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/cost/CostCalculatorModule.java]: I was going to do something else. This is no longer needed. I'm dropping this commit.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java [hunk 7]
```diff
@@ -147,9 +151,24 @@ private static boolean mustReplicate(JoinNode joinNode, Context context)
         return isAtMostScalar(joinNode.getRight(), context.getLookup());
     }
 
-    private static PlanNodeWithCost getJoinNodeWithCost(CostProvider costProvider, JoinNode possibleJoinNode)
+    private PlanNodeWithCost getJoinNodeWithCost(Context context, JoinNode possibleJoinNode)
     {
-        return new PlanNodeWithCost(costProvider.getCumulativeCost(possibleJoinNode), possibleJoinNode);
+        TypeProvider types = context.getSymbolAllocator().getTypes();
+        StatsProvider stats = context.getStatsProvider();
+        boolean replicated = possibleJoinNode.getDistributionType().get().equals(REPLICATED);
```
[the remark this answers is not in the document]
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java]: We call it `replicated` anywhere else. Let's keep it for now.

File: presto-main/src/main/java/com/facebook/presto/cost/ExchangeCostCalculator.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/cost/ExchangeCostCalculator.java]: After an offline discussion we decided to extract the `TaskCountEstimator` class

File: presto-main/src/main/java/com/facebook/presto/cost/ExchangeCostCalculator.java [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+    {
+        int numberOfNodesMultiplier = replicated ? numberOfNodes.getAsInt() : 1;
+
+        PlanNodeStatsEstimate probeStats = stats.getStats(probe);
+        PlanNodeStatsEstimate buildStats = stats.getStats(build);
+
+        double buildSideSize = buildStats.getOutputSizeInBytes(build.getOutputSymbols(), types);
+        double probeSideSize = probeStats.getOutputSizeInBytes(probe.getOutputSymbols(), types);
+
+        double cpuCost = probeSideSize + buildSideSize * numberOfNodesMultiplier;
```
[the remark this answers is not in the document]
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/cost/ExchangeCostCalculator.java]: https://github.com/prestodb/presto/pull/11982#discussion_r237305488

File: presto-main/src/main/java/com/facebook/presto/cost/ExchangeCostCalculator.java [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+            return probeCost
+                    .add(buildRemoteRepartitionCost)
+                    .add(buildLocalRepartitionCost);
+        }
+    }
+
+    /**
+     * Computes costs of local exchanges required to perform a join
+     */
+    public PlanNodeCostEstimate calculateJoinInputCost(
```
[the remark this answers is not in the document]
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/cost/ExchangeCostCalculator.java]: https://github.com/prestodb/presto/pull/11982#discussion_r237305488

File: presto-main/src/main/java/com/facebook/presto/server/CoordinatorModule.java (hunk 2 above)
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/server/CoordinatorModule.java]: I'm dropping this commit. It is not needed for what i'm doing now.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java (hunk 5 above)
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: Based on the offline discussion with @rschlussel it was decided that the `ExchangeCostCalculator` is a bad abstraction. Instead we introduced the `TaskCountEstimator`.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestDetermineJoinDistributionType.java [hunk 10]
```diff
@@ -238,6 +240,40 @@ public void testFlipAndReplicateWhenOneTableMuchSmaller()
                         values(ImmutableMap.of("A1", 0))));
     }
 
+    @Test
+    public void testFlipAndReplicateWhenOneTableMuchSmallerAndJoinCardinalityUnknown()
```
Comment (sopel39) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestDetermineJoinDistributionType.java]: Add similar test for `flip and repartition`, e.g: basic case when one table is smaller, but only like 2x and not 100x?

File: presto-main/src/main/java/com/facebook/presto/cost/TaskCountEstimator.java [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+                    .count());
+        };
+    }
+
+    public TaskCountEstimator(IntSupplier numberOfNodes)
+    {
+        this.numberOfNodes = requireNonNull(numberOfNodes, "numberOfNodes is null");
+    }
+
+    public int estimateSourceDistributedTaskCount()
```
[the remark this answers is not in the document]
Comment (rschlussel) [reply presto-main/src/main/java/com/facebook/presto/cost/TaskCountEstimator.java]: A partitioned exchange is determined by the `query.initial-hash-partitions` config property / `hash_partitions_count` session property.  At fb we set that to less than the number of nodes.

[the remark this answers is not in the document]
Comment (sopel39) [reply presto-main/src/main/java/com/facebook/presto/cost/TaskCountEstimator.java]: So if cluster has 2000 nodes then `query.initial-hash-partitions=2000`?

[the remark this answers is not in the document]
Comment (rschlussel) [reply presto-main/src/main/java/com/facebook/presto/cost/TaskCountEstimator.java]: We set it less than the number of nodes. e.g. if it was set to 500, the task count would be 500 after a partitioned exchange even though there are 2000 nodes.

The default value for the property is 100, so if you don't set the property and you have <100 nodes in your cluster, you won't notice it.  The number of hash partitions is min(number of nodes, hash_partition_count).

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java [hunk 12]
```diff
@@ -146,7 +147,8 @@ public PlanOptimizers(
             StatsCalculator statsCalculator,
             CostCalculator costCalculator,
             @EstimatedExchanges CostCalculator estimatedExchangesCostCalculator,
-            CostComparator costComparator)
+            CostComparator costComparator,
+            TaskCountEstimator taskCountEstimator)
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: I don't like this extra dependency here. This class is becoming a grab bag of everything someone might ever need.

Why is taskCountEstimator not an implementation detail of CostCalculator?

Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: `CostCalculator` is an interface that has 2 different implementation. It hash a single method, `calculateCost`, that computes cost of the entire operation. For joins this is the cost of exchanges needed for Join, cost of the Join itself, and cost of producing output out of Join. If the join cardinality cannot be estimated - the total cost of join cannot be estimated as well. What we are trying to achive here, is to be able to estimate cost of Join exchanges + cost of Join input processing. So even if the total join cost cannot be estimated, the sides still can be swapped.

Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: >  2 different implementation

One implementation works when exchanges were already introduced. The other one assumes there no exchange nodes introduced, so it adds the estimated cost of exchanges.

File: presto-main/src/main/java/com/facebook/presto/cost/TaskCountEstimator.java [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+{
+    private final IntSupplier numberOfNodes;
+
+    @Inject
+    public TaskCountEstimator(NodeSchedulerConfig nodeSchedulerConfig, InternalNodeManager nodeManager)
+    {
+        requireNonNull(nodeSchedulerConfig, "nodeSchedulerConfig is null");
+        requireNonNull(nodeManager, "nodeManager is null");
+        this.numberOfNodes = () -> {
+            Set<Node> activeNodes = nodeManager.getAllNodes().getActiveNodes();
```
[the remark this answers is not in the document]
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/cost/TaskCountEstimator.java]: It is gonna be accounted in a separate method, `estimatePartitionDistributedTaskCount`. For now we don't need this method, so this class only estimates `sourceDistributedTaskCount`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java (hunk 12 above)
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: As an alternative, i can add aditional non static methods specifically to the `CostCalculatorWithEstimatedExchanges`, and use that class directly (not by the `CostCalculator` interface) in the `DetermineJoinDistributionType`.

@martint What do you thin?

CC: @rschlussel @findepi @sopel39

Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: As i noted in https://github.com/prestodb/presto/pull/11982#discussion_r237281110:
> Cost calculators are passed to Rules from IterativeOptimizer, via the Context

if you want to change this design, we can talk about this. However, if this stays, then i think it should remain `IterativeOptimizer`'s responsibility to provide contextual information that rules need.

OTOH, would it work to implement `DetermineJoinDistributionType` in a simpler manner, based on stats calculator alone and not considering contextual information like number of nodes? I know considering more is supposed to yield better plans, but here we're in "rescue mode" anyway. Basing plan on stats alone (bypassing cost calculator) is not the route we want to take frequently.
@arhimondr would that work?

Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: > would it work to implement DetermineJoinDistributionType in a simpler manner, based on stats calculator alone and not considering contextual information like number of nodes?

It is easy to implement join sides flipping based on stats. However it is not that easy to implement replicated join selection based on stats. As replicated over repartitioned join requires different resources (more memory, more network, less cpu), and it is important to apply the cost weights there.

> but here we're in "rescue mode" anyway

For us this is gonna be a very common situation. Unfortunately we don't have NDV's for many tables. And without NDV it is not possible to estimate the JOIN output.

Comment (sopel39) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: Could you use `CostCalculatorUsingExchanges` calculator to calculate cost of join source plans with exchanges (created for the purpose of comparison) added? I think you did something similar for `UnionNode`.

Then you won't need to use directly `TaskCountEstimator`?

File: presto-main/src/main/java/com/facebook/presto/cost/TaskCountEstimator.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (sopel39) [reply presto-main/src/main/java/com/facebook/presto/cost/TaskCountEstimator.java]: Isn't source distributed task count governed by `query.initial-hash-partitions` too?

[the remark this answers is not in the document]
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/cost/TaskCountEstimator.java]: No. Currently source distributed stage may span across all the cluster nodes.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java (hunk 12 above)
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: > Could you use CostCalculatorUsingExchanges calculator to calculate cost of join source plans with synthetic exchange

Based on the offline discussion we agreed that it is not possible. As CostCalculatorUsingExchanges estimates the total cost of a JOIN, which cannot be estimated if the output JOIN cardinality cannot be estimated.

File: presto-main/src/main/java/com/facebook/presto/cost/CostCalculatorWithEstimatedExchanges.java [hunk 14]
```diff
@@ -199,6 +201,22 @@ private PlanNodeCostEstimate calculateJoinCost(PlanNode probe, PlanNode build, b
...
         }
 
+        @Override
+        public PlanNodeCostEstimate visitUnion(UnionNode node, Void context)
+        {
+            // this assumes that all union inputs will be gathered over the network
+            // that is not aways true
+            // but this estimate is better that returning UNKNOWN, as it sets
+            // cumulative cost to unknown
+            return CostCalculatorUsingExchanges.calculateExchangeCost(
```
[the remark this answers is not in the document]
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/cost/CostCalculatorWithEstimatedExchanges.java]: This is a problem with our cost mode. We don't know what type of exchange is gonna be introduced between the nodes, before these exchanges are actually introduced. I agree that this estimation will be wrong in some cases, but still i think it is better to propagate some cost than to return UNKNOWN. We never reorder UNION's with anything else. And for joins a miss-estimation of a UNION cost shouldn't matter much.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestDetermineJoinDistributionType.java (hunk 10 above)
Comment (arhimondr) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestDetermineJoinDistributionType.java]: The main reason for additional tests is to verify that the cost based decision can be made even when output size of a join cannot be estimated. Do you think it is worth adding yet another test for that purpose?

**Pull request merged**